### PR TITLE
add IAM AWS credential provider

### PR DESCRIPTION
**Pull request by balamurugana**

### CODE REVIEW

File: api/src/main/java/io/minio/credentials/IamAwsProvider.java [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+  }
+
+  @Override
+  public synchronized Credentials fetch() {
+    if (credentials != null && !credentials.isExpired()) {
+      return credentials;
+    }
+
+    HttpUrl url = this.customEndpoint;
+    if (getProperty("AWS_WEB_IDENTITY_TOKEN_FILE") != null) {
```
Comment (anjalshireesh) [inline api/src/main/java/io/minio/credentials/IamAwsProvider.java:113]: To making this call twice with same argument, we can save the output of this`getProperty` in a variable `tokenFile` and pass it as argument to `getToken`

File: api/src/main/java/io/minio/credentials/IamAwsProvider.java [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+          new WebIdentityProvider(
+              () -> getToken(),
+              url.toString(),
+              null,
+              null,
+              getProperty("AWS_ROLE_ARN"),
+              getProperty("AWS_ROLE_SESSION_NAME"),
+              httpClient);
+      credentials = provider.fetch();
+      return credentials;
```
Comment (anjalshireesh) [inline api/src/main/java/io/minio/credentials/IamAwsProvider.java:133]: Can be extracted out into a method `fetchUsingAwsTokenFile`

File: api/src/main/java/io/minio/credentials/IamAwsProvider.java [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+      if (url == null) {
+        url = HttpUrl.parse("http://169.254.169.254/latest/meta-data/iam/security-credentials/");
+      } else {
+        url =
+            new HttpUrl.Builder()
+                .scheme(url.scheme())
+                .host(url.host())
+                .addPathSegments("latest/meta-data/iam/security-credentials/")
+                .build();
+      }
```
Comment (anjalshireesh) [inline api/src/main/java/io/minio/credentials/IamAwsProvider.java:160]: Can be extracted into a method `getAwsCredentialsUrl`

File: api/src/main/java/io/minio/credentials/IamAwsProvider.java [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+        }
+
+        roleNames = response.body().string().split("\\R");
+      } catch (IOException e) {
+        throw new IllegalStateException("Unable to parse response", e);
+      }
+
+      if (roleNames.length == 0) {
+        throw new IllegalStateException("No IAM roles attached to EC2 service " + url);
+      }
```
Comment (anjalshireesh) [inline api/src/main/java/io/minio/credentials/IamAwsProvider.java:178]: Can be extracted into a method `getAwsRoleName`

File: api/src/main/java/io/minio/credentials/IamAwsProvider.java [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+    this.mapper = new ObjectMapper();
+    this.mapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
+    this.mapper.configure(MapperFeature.ACCEPT_CASE_INSENSITIVE_PROPERTIES, true);
+  }
+
+  private void checkLoopbackHost(HttpUrl url) {
+    try {
+      for (InetAddress addr : InetAddress.getAllByName(url.host())) {
+        if (!addr.isLoopbackAddress()) {
+          throw new IllegalArgumentException(url.host() + " is not loopback only host");
```
Comment (anjalshireesh) [inline api/src/main/java/io/minio/credentials/IamAwsProvider.java:71]: Not sure if `IllegalArgumentException` is the right choice for all the exceptions in this class.

Comment (anjalshireesh) [reply api/src/main/java/io/minio/credentials/IamAwsProvider.java]: Sorry I wrote this comment at the wrong place with the wrong exception name! I meant to say that I'm not sure if `IllegalStateException` makes sense for most of the exceptions in the class..

**Review by anjalshireesh — APPROVED** [review]
LGTM

**Review by sinhaashish — APPROVED** [review]
LGTM

**Pull request merged**